### PR TITLE
[SDPAP-7925]add config files

### PR DESCRIPTION
**Pull request by yeniatencio**
https://digital-vic.atlassian.net/browse/SDPAP-7925

  1.  Creates vocabulary if doesn't exist.
  2. Update field_publication_authors to use new vocabulary (Organisation) .
  3. Update cardinality.
  4. Update help text field.
BE link: https://nginx-php.pr-323.content-reference-sdp-vic-gov-au.sdp4.sdp.vic.gov.au/

Screenshot 2023-08-30 at 10 57 57 am

### CODE REVIEW

File: config/install/field.field.node.publication.field_publication_authors.yml [hunk 1]
```diff
@@ -1,16 +1,19 @@
+uuid: 53e1711e-9e06-497f-8b29-7361d0bd0273
```
Comment (MdNadimHossain) [inline config/install/field.field.node.publication.field_publication_authors.yml]: please don't add uuid for the config/install. This uuid represents the unique id for a specific site, so installing it for different project will create issues.

File: config/install/field.field.node.publication.field_publication_authors.yml [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 langcode: en
 status: true
 dependencies:
   config:
     - field.storage.node.field_publication_authors
     - node.type.publication
-    - taxonomy.vocabulary.department
+    - taxonomy.vocabulary.organisation
+_core:
+  default_config_hash: '-ExNyXJOfyV6z82S2g3GreY6Vnqw4F48Y1fJ5BVTRWw'
```
Comment (MdNadimHossain) [inline config/install/field.field.node.publication.field_publication_authors.yml]: this one as well, remove it. That gets generated when it exported for a specific site.

File: config/install/field.storage.node.field_publication_authors.yml [hunk 3]
```diff
@@ -1,9 +1,16 @@
+uuid: d0021eb1-fa3f-4711-a1d0-2a2399d6c0b5
```
Comment (MdNadimHossain) [inline config/install/field.storage.node.field_publication_authors.yml]: please remove this

File: config/install/field.storage.node.field_publication_authors.yml [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
 status: true
 dependencies:
   module:
+    - field_permissions
     - node
     - taxonomy
+third_party_settings:
+  field_permissions:
+    permission_type: public
+_core:
```
Comment (MdNadimHossain) [inline config/install/field.storage.node.field_publication_authors.yml]: please remove this

File: src/TidePublicationOperation.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+  /**
+   * Update field_publication_authors to use new vocabulary.
+   *
+   * @param \Drupal\Core\Config\Config $config
+   *   The config object.
+   * @param string $vocabulary
+   *   The new vocabulary.
+   */
+  public function updateFieldPublicationAuthorsVocabulary(Config $config, string $vocabulary) {
```
Comment (MdNadimHossain) [inline src/TidePublicationOperation.php]: why are we separating this like a helper function ? This needs to be used only once, after that this code will just sit here. Better to put it in update hook, cause after that there is no use of it. Also the values you set are constant, that also means this function can not be used for any other fields.

File: src/TidePublicationOperation.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }
+
+  /**
+   * Creates a new vocabulary.
+   *
+   * @param string $vocabulary
+   *   The new vocabulary.
+   */
+  public function createVocabulary(string $vocabulary) {
```
Comment (MdNadimHossain) [inline src/TidePublicationOperation.php:18]: This helper function we can keep, cause this can be used for creating any other vocabulary.

**Review by MdNadimHossain — CHANGES_REQUESTED** [review]
@yeniatencio please look into the comments. Also please add some behat test for this field if there isn't any.

File: src/TidePublicationOperation.php (hunk 5 above)
Comment (yeniatencio) [reply src/TidePublicationOperation.php]: Hey @MdNadimHossain , I am following Matthew's suggestions from this [pr](https://github.com/dpc-sdp/tide_core/pull/417#discussion_r1296544317) 
@krakerag, @MdNadimHossain  let me know guys how should I proceed then.

File: config/install/field.field.node.publication.field_publication_authors.yml (hunk 1 above)
Comment (yeniatencio) [reply config/install/field.field.node.publication.field_publication_authors.yml]: Thanks @MdNadimHossain , my bad, I didn't realise it was added it. I have removed them. Thanks.

**Comment by yeniatencio** [comment]
> @yeniatencio please look into the comments. Also please add some behat test for this field if there isn't any.

Hey @MdNadimHossain , Publication author field is not a new field and it is already contemplated in this behat test https://github.com/dpc-sdp/tide_publication/blob/a419a42ffb21f11ebdc6a22021890563b2101776/tests/behat/features/fields.feature#L36

File: src/TidePublicationOperation.php (hunk 5 above)
Comment (MdNadimHossain) [reply src/TidePublicationOperation.php]: I don't think this applies here, cause this function not going to be used after updating this field. Also for fresh install in the new CMS the update the updates will get in from the config/install for these fields. If there are scope where you can reuse this function then it make sense to separate it in a class. But this one not needed in the install hook and no other use case of reusing it as well.
CC: @krakerag

Comment (krakerag) [reply src/TidePublicationOperation.php]: This is an interesting one. All update hook code is throwaway code, but we established a pattern in tide_search and tide_core to minimize the code in the .install file, for the preference of putting code in a Helper or Operation class that lives in the src folder.

I think @yeniatencio is just following that established pattern here.

That said, I am happy to see either - I don't feel like being strict on install file rules is going to gain anything from a code readability standpoint.

I would be more strict on the createVocabulary method being in a reusable code location which is good on L86 below.

File: src/TidePublicationOperation.php (hunk 6 above)
Comment (krakerag) [reply src/TidePublicationOperation.php:18]: This is a good example of a code reuse pattern, good one @yeniatencio

**Review by MdNadimHossain — APPROVED** [review]
@yeniatencio good work. Thanks a lot for making all those changes :)

**Review by krakerag — APPROVED** [review]

**Pull request merged**